test/message_test.dart

```diff
@@ -255,88 +252,206 @@ void main() {
+          expect(message.encoding, equals(LATIN1));
+          expect(
+              message.headers,
+              containsPair('content-type',
+                  'application/octet-stream; charset=iso-8859-1'));
+        });
+
+        test('a non-ASCII body', () {
+          var message = _createMessage(body: _nonAscii, encoding: LATIN1);
+          expect(message.encoding, equals(LATIN1));
+          expect(
+              message.headers,
+              containsPair('content-type',
+                  'application/octet-stream; charset=iso-8859-1'));
+          expect(message.readAsBytes(), completion(equals(_latin1Bytes)));
+        });
+
+        test('body bytes', () {
+          var message = _createMessage(encoding: LATIN1, body: _latin1Bytes);
+          expect(message.encoding, equals(LATIN1));
+          expect(
+              message.headers,
+              containsPair('content-type',
+                  'application/octet-stream; charset=iso-8859-1'));
+          expect(message.readAsString(), completion(equals(_nonAscii)));
+        });
+      });
 
-    test("defaults to encoding a String as UTF-8", () {
-      expect(
-          _createMessage(body: "è").read().toList(),
-          completion(equals([
-            [195, 168]
-          ])));
-    });
+      test('a content-type header without a charset', () {
+        var message = _createMessage(
+            encoding: LATIN1, headers: {'Content-Type': 'text/plain'});
+        expect(message.headers,
+            containsPair('content-type', 'text/plain; charset=iso-8859-1'));
+      });
 
-    test("uses the explicit encoding if available", () {
-      expect(
-          _createMessage(body: "è", encoding: LATIN1).read().toList(),
-          completion(equals([
-            [232]
-          ])));
+      group('a content-type header and', () {
+        test('no body', () {
+          var message = _createMessage(
+              encoding: LATIN1,
+              headers: {'Content-Type': 'text/plain; charset=utf-8'});
+          expect(message.encoding, equals(LATIN1));
+          expect(message.headers,
+              containsPair('content-type', 'text/plain; charset=iso-8859-1'));
+        });
+
+        test('a plain ASCII body', () {
+          var message = _createMessage(
+              body: "foo",
+              encoding: LATIN1,
+              headers: {'Content-Type': 'text/plain; charset=utf-8'});
+          expect(message.encoding, equals(LATIN1));
+          expect(message.headers,
+              containsPair('content-type', 'text/plain; charset=iso-8859-1'));
+        });
+
+        test('a non-ASCII body', () {
+          var message = _createMessage(
+              body: _nonAscii,
+              encoding: LATIN1,
+              headers: {'Content-Type': 'text/plain; charset=utf-8'});
+          expect(message.encoding, equals(LATIN1));
+          expect(message.headers,
+              containsPair('content-type', 'text/plain; charset=iso-8859-1'));
```

Comment:
Do we care that this is the default charset assumed by the HTTP standard?

Reply:
No. It's possible we should have `readAsString()` default to Latin-1 instead of UTF-8, but I think matching `shelf`'s behavior is more important and I'm not eager to make a breaking change there.